cognite/client/data_classes/data_modeling/instances.py

```diff
@@ -0,0 +1,486 @@
+    source: ContainerId | ViewId
+    properties: dict[str, PropertyValue]
+
+    @classmethod
+    def load(cls, data: dict) -> NodeOrEdgeData:
+        return cls(**convert_all_keys_to_snake_case(data))
+
+    def dump(self, camel_case: bool = False) -> dict:
+        output = asdict(self)
+        if self.source:
+            output["source"] = self.source.dump(camel_case)
+        return convert_all_keys_to_camel_case_recursive(output) if camel_case else output
+
+
+class InstanceCore(DataModelingResource):
+    """A node or edge
+    Args:
+        instance_type (Literal["node", "edge"]) The type of instance.
+        space (str): The workspace for the instance.a unique identifier for the space.
+        external_id (str): Combined with the space is the unique identifier of the instance.
+    """
+
+    def __init__(self, space: str, external_id: str, instance_type: Literal["node", "edge"] = "node", **_: dict):
+        validate_data_modeling_identifier(space, external_id)
+        self.instance_type = instance_type
+        self.space = space
+        self.external_id = external_id
+
+
+class InstanceApply(InstanceCore):
+    """A node or edge. This is the write version of the instance.
+
+    Args:
+        instance_type (Literal["node", "edge"]) The type of instance.
+        space (str): The workspace for the instance.a unique identifier for the space.
+        external_id (str): Combined with the space is the unique identifier of the instance.
+        existing_version (int): Fail the ingestion request if the node's version is greater than or equal to this value.
+                                If no existingVersion is specified, the ingestion will always overwrite any
+                                existing data for the edge (for the specified container or instance). If existingVersion is
+                                set to 0, the upsert will behave as an insert, so it will fail the bulk if the
+                                item already exists. If skipOnVersionConflict is set on the ingestion request,
+                                then the item will be skipped instead of failing the ingestion request.
+        sources (list[NodeOrEdgeData]): List of source properties to write. The properties are from the instance and/or
+                        container the container(s) making up this node.
+    """
+
+    def __init__(
+        self,
+        space: str,
+        external_id: str,
+        instance_type: Literal["node", "edge"] = "node",
+        existing_version: int = None,
+        sources: list[NodeOrEdgeData] = None,
+        **_: dict,
+    ):
+        super().__init__(space, external_id, instance_type)
+        self.existing_version = existing_version
+        self.sources = sources
+
+    def dump(self, camel_case: bool = False) -> dict[str, Any]:
+        output = super().dump(camel_case)
+        if self.sources:
+            output["sources"] = [source.dump(camel_case) for source in self.sources]
```

Comment:
Should also include "existing_version"?

Reply:
Ah right, the implementation in the super class already handles it. Never mind then 👍